packages/tecrock-simulation/src/storage.ts

```diff
@@ -16,33 +18,106 @@ export function initDatabase() {
-  const dbRef = ref(db, "models/" + modelId);
-
-  onValue(dbRef, (data) => {
-    callback(data.val().model);
-  }, {
-    onlyOnce: true
+  const url = `https://models-resources.concord.org/te-models/${modelId}/model.json.gz`;
+
+  console.info("📦 Attempting to load model from S3...");
+
+  fetch(url)
+  .then((response) => {
+    if (response.ok) {
+      console.info("📦 Loaded model from S3.");
+      return response.json();
+    } else {
+      console.info("🔥 Falling back to loading model from Firebase...");
+      return fallbackToLoadFromFirebase(modelId);
+    }
+  })
+  .then((data) => {
+    callback(data);
+  })
+  .catch((error) => {
+    showError("Failed to load model from the cloud.", error);
   });
 }
+
+// THIS WILL BE REMOVED IN PHASE 2
+function fallbackToLoadFromFirebase(modelId: string): Promise<any> {
+  return new Promise<any>((resolve, reject) => {
+    if (!app) {
+      reject("Firebase app is not initialized.");
+      return;
+    }
+    const db = getDatabase(app);
+    const dbRef = ref(db, "models/" + modelId);
+
+    onValue(dbRef, (data) => {
+      if (data.exists()) {
+        console.info("🔥 Loaded model from Firebase.");
+        resolve(data.val().model);
+      } else {
+        reject("Model not found.");
+      }
+    }, {
+      onlyOnce: true
+    })
+  });
+}
+
+async function saveCompressedModelToS3(serializedModel: ISerializedState): Promise<string> {
+  return new Promise<string>(async (resolve, reject) => {
+    try {
+      const client = new TokenServiceClient({ env: "production" });
+      const filename = "model.json.gz";
+      const resource: S3Resource = await client.createResource({
+        tool: "te-models",
+        type: "s3Folder",
+        name: filename,
+        description: "Created by TE simulation",
+        accessRuleType: "readWriteToken"
+      }) as S3Resource;
+      const readWriteToken = client.getReadWriteToken(resource) || "";
+      const credentials = await client.getCredentials(resource.id, readWriteToken);
+
+      // S3 configuration is based both on resource and credentials info.
+      const { bucket, region } = resource;
+      const { accessKeyId, secretAccessKey, sessionToken } = credentials;
+      const s3 = new S3({ region, accessKeyId, secretAccessKey, sessionToken });
+      const publicPath = client.getPublicS3Path(resource, filename);
+
+      // Compress the model using pako (Gzip) into a blob for uploading
+      const data = JSON.stringify(serializedModel);
+      const compressedData = pako.gzip(data);
+      const blob = new Blob([compressedData], { type: "application/gzip" });
+
+      await s3.upload({
+        Bucket: bucket,
+        Key: publicPath,
+        Body: blob,
+        ContentType: "application/json",
+        ContentEncoding: "gzip",
+        CacheControl: "public, max-age=31536000, immutable", // Cache forever (1 year)
+      }).promise();
+
+      const [_folder, id, _filename] = publicPath.split("/");
```

Comment:
Looks like you could get the id directly with `resource.id` instead of parsing the path which might change via a configuration change within the token service.

Reply:
Good point.  I'll update the code.

Reply:
Done in https://github.com/concord-consortium/tectonic-explorer/pull/365/commits/b1c10dc9838b17cd21b9943eecdd3984fa4edbe6